pkg/roachprod/install/expander.go

```diff
@@ -329,3 +333,40 @@ func (e *expander) maybeExpandCertsDir(
 	}
 	return c.CertsDir(e.node), true, nil
 }
+
+// maybeExpandIPAddress is an expanderFunc for {ipaddress:<nodeSpec>[:public|private]}
+func (e *expander) maybeExpandIPAddress(
+	ctx context.Context, l *logger.Logger, c *SyncedCluster, cfg ExpanderConfig, s string,
+) (string, bool, error) {
+	m := ipAddressRe.FindStringSubmatch(s)
+	if m == nil {
+		return s, false, nil
+	}
+
+	var err error
+	switch m[2] {
+	case ":public":
+		if e.publicIPs == nil {
+			e.publicIPs = make(map[Node]string, len(c.VMs))
+			for _, node := range allNodes(len(c.VMs)) {
+				e.publicIPs[node] = c.Host(node)
+			}
+		}
+
+		s, err = e.maybeExpandMap(c, e.publicIPs, m[1])
+	default:
+		if e.privateIPs == nil {
+			e.privateIPs = make(map[Node]string, len(c.VMs))
+			for _, node := range allNodes(len(c.VMs)) {
+				ip, err := c.GetInternalIP(node)
+				if err != nil {
+					return "", false, err
+				}
+				e.privateIPs[node] = ip
+			}
+		}
+
+		s, err = e.maybeExpandMap(c, e.privateIPs, m[1])
```

Comment:
It looks like the current regex would match singletons, e.g., `{ip:1}`, or even empty ranges, `{ip:-}`. Both would yield out-of-bounds panic inside `maybeExpandMap`.

Reply:
>even empty ranges, {ip:-}
```
darrylwong@Darryls-MacBook-Pro-2 cockroach % ./bin/roachprod run local "echo {ip:-}" 
local:[1 2 3 4]: echo {ip:-} 0/4 
Error: UNCLASSIFIED_PROBLEM: unable to parse node selector '-': strconv.Atoi: parsing "": invalid syntax
(1) UNCLASSIFIED_PROBLEM
Wraps: (2) error expanding command: echo {ip:-}
Wraps: (3) attached stack trace
```

The error message isn't super useful, but it doesn't panic. I was mostly copying the existing behavior from pgurl/pgport, but I'll switch it to use a different regex that enforces `nodeSpec` correctness.

>would match singletons, e.g., {ip:1}

I'm not sure if I'm following what you mean by this. Matching for only one node is intended and handled by the `ListNodes` parsing. Let me know if I'm missing something.

Reply:
I see. Somehow I missed the fact that `nodeSpec` includes the `:` prefix; i.e., on a first reading, I thought the prefix wasn't captured. It's kind of a subtle implementation detail which guards `nodeSpec[1:]` (in `maybeExpandMap`).